File: packages/acr_oracle_client/acr_oracle_client/agentcard.py
"""The agent card — who is calling, signed, with no registry to ask.

An agent presents a card. The card says which key it is, what role it is acting
in, which service it is for, and until when. The gate recovers the signature and
believes the key, not a list — so nobody has to be enrolled, and there is no
registry to run, gatekeep, or lose.

THE DOMAIN OMITS `verifyingContract`, AND THAT IS WHAT PERMISSIONLESS MEANS HERE.
The four domains this project already has — `ACR Oracle`,
`ACR AttestationRegistry`, `ACR Receipt Mirror`, `ACR Human Id Mirror` — each bind
to a deployed address, because each is verified BY that contract. A card is
verified by whoever reads it, so there is no address to name.

That buys openness and costs two things, both paid for in the struct rather than
hoped away:

  * A `{name, version, chainId}` domain is valid at ANY verifier on that chain.
    So the card carries `audience`, and a gate refuses a card addressed elsewhere.
    Without it, a card minted for some other ACR-domain service would be
    presentable here — which is precisely the replay a `verifyingContract`
    normally prevents.

  * A bearer credential is replayable inside its validity window by anyone who
    captures it. So `MAX_TTL_S` bounds that window. There is no per-call nonce on
    purpose: an agent cannot afford a round trip before every request, and the
    AgentKit verifier in this same codebase made the same call — freshness lives
    inside the signed message rather than in a table on our side.

AND A CARD PROVES IDENTITY, NOT SCARCITY. Keys are free, so a caller who wants
ten identities can have them, and a rate limit keyed on the agent address is
evadable by anyone willing to run `eth-account` in a loop. That is why
`humanCluster` exists: a card may CLAIM a human, the gate checks that claim
against `HumanIdMirror.clusterOf` on chain, and the limit keys on the human. Ten
keys belonging to one person then buy one budget. The claim is worth nothing
unverified, which is why this module only carries it and `agentgate` is the thing
that checks it.
"""

from __future__ import annotations

import base64
import binascii
import json
import time
from dataclasses import dataclass, replace

#: `("ACR <Thing>", "1")` — the house idiom, and this was the free slot.
CARD_DOMAIN_NAME = "ACR Agent Card"
CARD_DOMAIN_VERSION = "1"

#: The roles a card may claim. Four come from `_ROLE_WALLET_FIELDS` in
#: `signer.py` so the card and the venue's service accounts share one vocabulary
#: rather than drifting into two; `reader` is the fifth, for the query path, which
#: has no wallet because it spends nothing.
ROLES: tuple[str, ...] = ("maker", "taker", "poster", "owner", "reader")

#: The longest life a card may claim. A bearer credential is replayable until it
#: expires, so this is the blast radius of a captured card and not a convenience
#: knob. Fifteen minutes is long enough for an agent to make many calls on one
#: signature and short enough that a leaked card is worth little.
MAX_TTL_S = 900

#: Tolerance for a signer whose clock runs ahead of ours. Without it, two honest
#: machines disagreeing by a second make a valid card look not-yet-valid — and
#: "your clock is fast" is an unhelpful 401.
CLOCK_SKEW_S = 30

ZERO32 = "0x" + "00" * 32

AGENT_CARD_TYPES: dict[str, list[dict[str, str]]] = {
    "AgentCard": [
        {"name": "agent", "type": "address"},
        {"name": "name", "type": "string"},
        {"name": "role", "type": "string"},
        {"name": "audience", "type": "string"},
        {"name": "scopeHash", "type": "bytes32"},
        {"name": "humanCluster", "type": "bytes32"},
        {"name": "issuedAt", "type": "uint64"},
        {"name": "expiresAt", "type": "uint64"},
    ]
}


def agentcard_domain(chain_id: int) -> dict:
    """The EIP-712 domain a card is signed under.

    No `verifyingContract`, deliberately — see the module docstring. `chainId`
    stays because a card minted for Arc should not be presentable on another
    chain's deployment of this same service.
    """
    return {
        "name": CARD_DOMAIN_NAME,
        "version": CARD_DOMAIN_VERSION,
        "chainId": int(chain_id),
    }


def scope_hash(scopes) -> str:
    """keccak of the SORTED, de-duplicated scope list.

    Sorted because `["read","write"]` and `["write","read"]` are the same grant
    and must not produce two different cards; de-duplicated for the same reason.
    An empty list hashes to the zero word rather than to keccak("") so "no scopes
    claimed" is distinguishable from "claimed an empty string".
    """
    from eth_utils import keccak

    items = sorted({str(s).strip() for s in (scopes or []) if str(s).strip()})
    if not items:
        return ZERO32
    return "0x" + keccak(text=",".join(items)).hex()


@dataclass(frozen=True)
class AgentCard:
    """What an agent asserts about itself. Frozen: a verified card must not be
    editable by the code that received it."""

    agent: str
    name: str
    role: str
    audience: str
    scope_hash: str = ZERO32
    #: A CLAIM, worth nothing until `agentgate` checks it on chain.
    human_cluster: str = ZERO32
    issued_at: int = 0
    expires_at: int = 0

    @property
    def claims_human(self) -> bool:
        return self.human_cluster.lower() != ZERO32

    @property
    def ttl_s(self) -> int:
        return int(self.expires_at) - int(self.issued_at)

    def message(self) -> dict:
        """The EIP-712 message. Field names are the struct's, not Python's."""
        from web3 import Web3

        return {
            "agent": Web3.to_checksum_address(self.agent),
            "name": self.name,
            "role": self.role,
            "audience": self.audience,
            "scopeHash": self.scope_hash,
            "humanCluster": self.human_cluster,
            "issuedAt": int(self.issued_at),
            "expiresAt": int(self.expires_at),
        }

    def to_json(self) -> dict:
        """Wire form. Snake_case, because this crosses into HTTP where the rest
        of this service's JSON is snake_case; `message()` is the only place the
        struct's camelCase names appear."""
        return {
            "agent": self.agent,
            "name": self.name,
            "role": self.role,
            "audience": self.audience,
            "scope_hash": self.scope_hash,
            "human_cluster": self.human_cluster,
            "issued_at": int(self.issued_at),
            "expires_at": int(self.expires_at),
        }

    @classmethod
    def from_json(cls, d: dict) -> AgentCard:
        return cls(
            agent=str(d["agent"]),
            name=str(d.get("name") or ""),
            role=str(d.get("role") or ""),
            audience=str(d.get("audience") or ""),
            scope_hash=str(d.get("scope_hash") or ZERO32),
            human_cluster=str(d.get("human_cluster") or ZERO32),
            issued_at=int(d.get("issued_at") or 0),
            expires_at=int(d.get("expires_at") or 0),
        )


def mint(
    agent: str,
    *,
    name: str,
    role: str,
    audience: str,
    scopes=None,
    human_cluster: str = ZERO32,
    ttl_s: int = 300,
    now: int | None = None,
) -> AgentCard:
    """A card with its window already set. Refuses a TTL over the bound here
    rather than letting a gate discover it later — the mistake is cheaper to fix
    at the signer than at the verifier."""
    if role not in ROLES:
        raise ValueError(f"unknown role {role!r}; expected one of {', '.join(ROLES)}")
    if ttl_s <= 0:
        raise ValueError("ttl_s must be positive — a card that expires on issue is not a card")
    if ttl_s > MAX_TTL_S:
        raise ValueError(f"ttl_s {ttl_s} exceeds MAX_TTL_S {MAX_TTL_S}")
    t = int(time.time()) if now is None else int(now)
    return AgentCard(
        agent=agent, name=name, role=role, audience=audience,
        scope_hash=scope_hash(scopes), human_cluster=human_cluster,
        issued_at=t, expires_at=t + int(ttl_s),
    )


def card_digest(card: AgentCard, chain_id: int) -> bytes:
    """The EIP-712 digest a signature covers.

    Exposed rather than kept private because `humanid.py` proves its own encoding
    against the contract's `resolutionDigest` view for exactly this reason: an
    encoding mistake recovers to a stranger, and the only symptom is a 401 on a
    call that should have worked. There is no contract to check against here, so
    the tests check this against `eth_account`'s own recovery instead.
    """
    from eth_account.messages import encode_typed_data
    from eth_utils import keccak

    signable = encode_typed_data(
        domain_data=agentcard_domain(chain_id),
        message_types=AGENT_CARD_TYPES,
        message_data=card.message(),
    )
    return bytes(keccak(b"\x19" + signable.version + signable.header + signable.body))


def sign_card(card: AgentCard, signer, chain_id: int) -> str:
    """Sign with anything satisfying the `Signer` protocol — raw key or Circle
    custody. Returns a 65-byte `0x` signature, which is what travels in a header;
    the `(v, r, s)` triple the protocol returns is a contract-call shape."""
    v, r, s = signer.sign_typed_data(
        agentcard_domain(chain_id), AGENT_CARD_TYPES, card.message(), "AgentCard"
    )
    return "0x" + bytes(r).hex() + bytes(s).hex() + bytes([int(v)]).hex()


def recover_agent(card: AgentCard, signature: str, chain_id: int) -> str:
    """The address that actually signed, checksummed.

    A caller must compare this to `card.agent` itself. Returning the recovered
    address rather than a boolean is deliberate: a function that answered
    "valid: true" would invite being trusted without anyone checking WHICH key it
    validated, and the card's entire claim is about which key.
    """
    from eth_account import Account
    from eth_account.messages import encode_typed_data

    signable = encode_typed_data(
        domain_data=agentcard_domain(chain_id),
        message_types=AGENT_CARD_TYPES,
        message_data=card.message(),
    )
    return str(Account.recover_message(signable, signature=signature))


# --- the wire format ---------------------------------------------------------
#
# base64(JSON), matching how x402 carries PAYMENT-SIGNATURE and how the human
# gate carries HUMAN-PROOF. One encoding across three gates means an agent author
# writes one decoder.

def encode_header(card: AgentCard, signature: str) -> str:
    raw = json.dumps({"card": card.to_json(), "signature": signature},
                     separators=(",", ":"), sort_keys=True).encode()
    return base64.b64encode(raw).decode()


def decode_header(value: str) -> tuple[AgentCard, str]:
    """Parse a presented header.

    Raises `ValueError` with a reason on anything malformed, so a gate can answer
    401 with something an agent author can act on. Bare `Exception` would collapse
    "you sent nonsense" and "we broke" into one message.
    """
    try:
        payload = json.loads(base64.b64decode(value, validate=True))
    except (binascii.Error, ValueError, TypeError) as exc:
        raise ValueError("header is not base64-encoded JSON") from exc
    if not isinstance(payload, dict):
        raise ValueError("header did not decode to an object")
    card_raw, sig = payload.get("card"), payload.get("signature")
    if not isinstance(card_raw, dict):
        raise ValueError("header is missing the card object")
    if not isinstance(sig, str) or not sig.startswith("0x"):
        raise ValueError("header is missing a 0x signature")
    try:
        return AgentCard.from_json(card_raw), sig
    except (KeyError, TypeError, ValueError) as exc:
        raise ValueError(f"card is missing or malformed: {exc}") from exc


def with_window(card: AgentCard, **kw) -> AgentCard:
    """A copy with fields replaced — used by tests to build the near-miss cases
    without hand-assembling eight fields each time."""
    return replace(card, **kw)

File: apps/terminal/lib/agentcard.ts
/* The agent card, minted where a developer can watch it happen.
 *
 * Isomorphic on purpose. The same function signs a visitor's THROWAWAY card in
 * their browser tab and the demo human's card inside the probe route on the
 * server — two callers, one encoder, so the two cards cannot disagree about the
 * struct. It mirrors `packages/acr_oracle_client/acr_oracle_client/agentcard.py`
 * field for field; the buyer agent's identical encoder is pinned against the Python
 * gate in `tests/test_agent_card_cross_language.py`, and `agentcard.test.ts` pins
 * this one's shape against that.
 *
 * WHY THE DOMAIN NAMES NO CONTRACT. Every other EIP-712 domain in this project
 * binds to a deployed address because a contract verifies it. A card is verified by
 * whoever reads it, so there is nothing to name — and that is what permissionless
 * means here. The cost is paid inside the struct: `audience` stands in for the
 * missing `verifyingContract`, and a 15-minute bound replaces a nonce.
 *
 * viem is imported lazily because the only client that needs it is one button on
 * /developers, and the page's first paint should not carry a signing library.
 */

/** The header the gate reads — `CARD_HEADER` in agentgate.py. */
export const CARD_HEADER = "AGENT-CARD";

export const DOMAIN_NAME = "ACR Agent Card";
export const DOMAIN_VERSION = "1";
export const ZERO32 = `0x${"00".repeat(32)}` as const;

/** `MAX_TTL_S` in agentcard.py. A longer card is refused by the gate, so minting one
 *  is a wasted round trip; clamped here rather than rejected. */
export const MAX_TTL_S = 900;

export const ROLES = ["maker", "taker", "poster", "owner", "reader"] as const;
export type Role = (typeof ROLES)[number];

/** The struct, in the order the Python side hashes it. */
export const AGENT_CARD_TYPES = {
  AgentCard: [
    { name: "agent", type: "address" },
    { name: "name", type: "string" },
    { name: "role", type: "string" },
    { name: "audience", type: "string" },
    { name: "scopeHash", type: "bytes32" },
    { name: "humanCluster", type: "bytes32" },
    { name: "issuedAt", type: "uint64" },
    { name: "expiresAt", type: "uint64" },
  ],
} as const;

export interface MintOptions {
  privateKey: `0x${string}`;
  chainId: number;
  /** Who the card is FOR. Read it from `/agent/challenge`, never hardcode it. */
  audience: string;
  name?: string;
  role?: Role;
  ttlSeconds?: number;
  /** A human cluster to CLAIM. Unset claims nothing. The gate refuses a claim it
   *  cannot confirm on chain with a 401, so this is never defaulted. */
  humanCluster?: `0x${string}`;
}

export interface MintedCard {
  /** The base64 header value — what goes in `AGENT-CARD`. */
  header: string;
  /** The signing address, so the UI can show WHO just got a tier. */
  agent: `0x${string}`;
  expiresAt: number;
  claimsHuman: boolean;
}

/** A fresh throwaway key. Lives in memory for as long as the caller keeps it. */
export async function throwawayKey(): Promise<`0x${string}`> {
  const { generatePrivateKey } = await import("viem/accounts");
  return generatePrivateKey();
}

/** Mint, sign, encode. */
export async function mintCard(opts: MintOptions): Promise<MintedCard> {
  const { privateKeyToAccount } = await import("viem/accounts");
  const ttl = Math.min(Math.max(1, Math.floor(opts.ttlSeconds ?? 300)), MAX_TTL_S);
  const account = privateKeyToAccount(opts.privateKey);
  const issuedAt = Math.floor(Date.now() / 1000);
  const expiresAt = issuedAt + ttl;

  const message = {
    agent: account.address,
    name: opts.name ?? "acr-developers-page",
    role: opts.role ?? "reader",
    audience: opts.audience,
    scopeHash: ZERO32,
    humanCluster: opts.humanCluster ?? ZERO32,
    issuedAt: BigInt(issuedAt),
    expiresAt: BigInt(expiresAt),
  };

  const signature = await account.signTypedData({
    domain: { name: DOMAIN_NAME, version: DOMAIN_VERSION, chainId: opts.chainId },
    types: AGENT_CARD_TYPES,
    primaryType: "AgentCard",
    message,
  });

  // snake_case on the wire, like `AgentCard.to_json`. Names matter, order does not;
  // sorted anyway so two encoders produce byte-identical headers for the same card.
  const card = {
    agent: message.agent,
    audience: message.audience,
    expires_at: expiresAt,
    human_cluster: message.humanCluster,
    issued_at: issuedAt,
    name: message.name,
    role: message.role,
    scope_hash: message.scopeHash,
  };
  // The JSON is ASCII (hex and plain strings), so `btoa` is safe in both runtimes
  // and `Buffer` — Node-only — is not needed.
  const header = btoa(JSON.stringify({ card, signature }));
  return { header, agent: account.address, expiresAt, claimsHuman: message.humanCluster !== ZERO32 };
}

/** The four demo humans' wallet labels, for the probe route's "as a demo human".
 *  Keys derive from these labels exactly as `demo_humans.py` derives them —
 *  `sha256("acr-buyer::" + label)` — which is public by construction. The probe
 *  route derives on the SERVER so no key ever reaches a browser; the browser only
 *  ever sees the resulting header. `acr-buyer-4` is the solo human: the one whose
 *  wallet has the least at stake, and whose cluster is visibly different from the
 *  fleet's three. */
export const DEMO_HUMAN_LABEL = "acr-buyer-4";

/** Derive a demo wallet's key from its label — server-side only. */
export async function demoKey(label: string): Promise<`0x${string}`> {
  const { sha256, toBytes } = await import("viem");
  return sha256(toBytes(`acr-buyer::${label}`));
}
